### DISCUSSION TITLE
Publish core frontend methods to NPM

### DISCUSSION THREAD

**Discussion by joshuabaker** [verbb/formie#1641] [Ideas]
### What are you trying to do?

It’d be great to have access to some of [the logic you’ve coded](https://github.com/verbb/formie/tree/1a63315fe39d672bb2e70b1e8a5e88d6e6c22b0b/src/web/assets/frontend) for the frontend in an NPM package. This would allow us to pull common logic into other implementations (e.g. React, Vue, et al.).

### What's your proposed solution?

Add a `package.json` in the frontend directory or, perhaps, split the frontend logic into its own repo and pull it into this one.

I would imagine some adjustment to the API surface is needed. Looking now, I don’t imagine this would be significant.

### Additional context

The primary reason for the ask is to allow us to better match your ongoing changes with our [React implementation](https://github.com/joshuabaker/react-formie-forms).

**Comment by engram-design** [comment]
Yep, this has been on my list for a while. I actually thought there was an open issue for this already, but I can't seem to find it.

But yes, I just need to give some more proper thought to building this out as a proper JS API that'll work well detached from the Formie library. There's a few things I need to consider

**Comment by joshuabaker** [comment]
@engram-design It’s quite likely that we’re going to abandon/archive the React implementation at some point this year.

To broaden the scope of this issue a bit, have you put any further consideration towards shipping client-side libraries as part of Formie?

**Comment by engram-design** [comment]
Did you mean to close this? Certainly on our list to do!

**Comment by joshuabaker** [comment]
@engram-design I’m abandoning my [React implementation](https://github.com/joshuabaker/react-formie-forms). 

Please can you convert this to an ‘Ideas’ discussion.

**Comment by engram-design** [comment]
Ah, gotcha. I may take over that similar to https://github.com/verbb/formie-headless, but regardless, we certainly want to provide an npm package for Formie to make things easier for devs!

**Reply by Anubarak** [reply]
@engram-design Hey Josh, were you already able to publish an npm package?
We would like to use formie in a vue SPA and would like to use your JS.

**Comment by joshuabaker** [comment]
@engram-design Bump. Josh, please can you convert this to an ‘Ideas’ discussion. I don’t have the ability to do that.

**Comment by engram-design** [comment]
Updated in [4.0.0-beta.3](https://github.com/verbb/formie/releases/tag/4.0.0-beta.3).

We've now got dedicated NPM packages, and a re-architecture to make Formie work better for modern JS, or BYO in your own bundle, with different framework (or none) support. Have a read through the [docs](https://verbb.io/craft-plugins/formie/docs/v4/frontend/frontend-assets).

Check out the dedicated front-end package [docs](https://docs.verbb.io/formie) too.
